central/complianceoperator/v2/report/manager/watcher/scanwatcher.go

```diff
@@ -0,0 +1,239 @@
+	Stop()
+}
+
+// ScanWatcherResults is returned when the watcher detects that the scan is completed.
+type ScanWatcherResults struct {
+	Ctx          context.Context
+	WatcherID    string
+	Scan         *storage.ComplianceOperatorScanV2
+	CheckResults set.StringSet
+	Error        error
+}
+
+// IsComplianceOperatorHealthy indicates whether Compliance Operator is ready for automatic reporting
+func IsComplianceOperatorHealthy(ctx context.Context, clusterID string, complianceIntegrationDataStore complianceIntegrationDS.DataStore) error {
+	coStatus, err := complianceIntegrationDataStore.GetComplianceIntegrationByCluster(ctx, clusterID)
+	if err != nil {
+		return errors.Wrap(err, ComplianceOperatorIntegrationDataStoreError.Error())
+	}
+	if len(coStatus) == 0 {
+		log.Errorf("No compliance integrations retrieved from cluster %s", clusterID)
+		return ComplianceOperatorIntegrationZeroIntegrationsError
+	}
+	if !coStatus[0].GetOperatorInstalled() {
+		return ComplianceOperatorNotInstalledError
+	}
+	if semver.Compare(coStatus[0].GetVersion(), minimumComplianceOperatorVersion) < 0 {
+		return ComplianceOperatorVersionError
+	}
+	return nil
+}
+
+// GetWatcherIDFromScan given a Scan, returns a unique ID for the watcher
+func GetWatcherIDFromScan(scan *storage.ComplianceOperatorScanV2) (string, error) {
+	clusterID := scan.GetClusterId()
+	if clusterID == "" {
+		return "", errors.New("Missing cluster ID")
+	}
+	scanID := scan.GetId()
+	if scanID == "" {
+		return "", errors.New("Missing scan ID")
+	}
+	startTime := scan.GetLastStartedTime()
+	if startTime == nil {
+		// This could happen if the scan is freshly created
+		log.Debug("Missing last stated time")
+		return "", nil
+	}
+	return fmt.Sprintf("%s:%s:%s", clusterID, scanID, startTime.String()), nil
+}
+
+// GetWatcherIDFromCheckResult given a CheckResult, returns a unique ID for the watcher
+func GetWatcherIDFromCheckResult(ctx context.Context, result *storage.ComplianceOperatorCheckResultV2, scanDataStore scanDS.DataStore) (string, error) {
+	scanRefQuery := search.NewQueryBuilder().AddExactMatches(
+		search.ComplianceOperatorScanRef,
+		result.GetScanRefId(),
+	).ProtoQuery()
+	scans, err := scanDataStore.SearchScans(ctx, scanRefQuery)
+	if err != nil {
+		return "", errors.Errorf("Unable to retrieve scan : %v", err)
+	}
+	if len(scans) == 0 {
+		return "", errors.Errorf("Scan not found for result %s", result.GetCheckName())
+	}
+	var startTime string
+	var ok bool
+	if startTime, ok = result.GetAnnotations()[lastScannedAnnotationKey]; !ok {
+		return "", errors.Errorf("%s annotation not found", lastScannedAnnotationKey)
+	}
+	timestamp, err := protocompat.ParseRFC3339NanoTimestamp(startTime)
+	if err != nil {
+		return "", errors.Errorf("Unable to parse time: %v", err)
+	}
+	if timestamp.String() != scans[0].GetLastStartedTime().String() {
+		return "", errors.New("The result and the scan do not have the same timestamp")
+	}
+	return GetWatcherIDFromScan(scans[0])
```

Comment:
We discussed it in the call, but I have another idea - we could actually use that block in a for-loop in case maybe the index `[1]` has the matching timestamp. That would save us a disruption and could maybe recover gracefully. Does that make sense?